src/wallet/wallet.cpp

```diff
@@ -1906,7 +1906,12 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         // Check if any destination contains a witness program:
         int witnessversion = 0;
         std::vector<unsigned char> witnessprogram;
-        if (recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
+        const bool is_wit = recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram);
+        if (is_wit && witnessversion >= 1 && GetScriptPubKeyMan(OutputType::BECH32M, true)) {
```

Comment:
4bf448fc2528d9df371b2ca6e9b13319441a21e5: I don't understand what the `witnessversion >= 1` is doing.

Assuming `witnessversion` is `2`, then a bech32m address will be selected (possibly a taproot address). However, if the `witnessversion` is `0`, even with a taproot wallet, a segwit_0 address will be enforced. This seems inconsistent and confusing, unless I am missing something?

Reply:
This opportunistic change code was initially intended to _upgrade_ change addresses to `bech32` if there was no privacy downside. With this new change it seems that we _downgrade_ a v1+ witness to v0 just in order to blend in. I'm also not sure if we should do that. It could actually hurt privacy when that v0 change address is spent later on and combined with a v1 input.

Reply:
Indeed. I've changed this to just return `BECH32M` or `BECH32` depending on what is available, with a preference for `BECH32M`.

However I think this function (and generally how change address type is determined) should be revisited because not all address types are available in descriptor wallets. But I think that is out of scope for this PR.